Sources/ASTGen/ASTGen.swift

```diff
@@ -0,0 +1,253 @@
+    
+    return elements.withBridgedArrayRef { elementsRef in
+      SwiftTupleExpr_create(self.ctx, lParenLoc, elementsRef, rParenLoc)
+    }
+  }
+
+  public func visit(_ node: VariableDeclSyntax) -> UnsafeMutableRawPointer {
+    let pattern = visit(node.bindings.first!.pattern)
+    let initializer = visit(node.bindings.first!.initializer!)
+    
+    let loc = self.base.advanced(by: node.position.utf8Offset).raw
+    let isStateic = false // TODO: compute this
+    let isLet = node.letOrVarKeyword.tokenKind == .letKeyword
+
+    // TODO: don't drop "initializer" on the floor.
+    return SwiftVarDecl_create(ctx, pattern, loc, isStateic, isLet, declContext)
+  }
+  
+  public func visit(_ node: ConditionElementSyntax) -> UnsafeMutableRawPointer {
+    visit(node.condition)
+  }
+
+  public func visit(_ node: CodeBlockItemSyntax) -> UnsafeMutableRawPointer {
+    visit(node.item)
+  }
+
+  public func visit(_ node: CodeBlockSyntax) -> UnsafeMutableRawPointer {
+    let statements = node.statements.map(self.visit)
+    let loc = self.base.advanced(by: node.position.utf8Offset).raw
+
+    return statements.withBridgedArrayRef { ref in
+      BraceStmt_create(ctx, loc, ref, loc)
+    }
+  }
+
+  public func visit(_ node: FunctionParameterSyntax) -> UnsafeMutableRawPointer {
+    let loc = self.base.advanced(by: node.position.utf8Offset).raw
+    
+    let firstName: UnsafeMutableRawPointer?
+    let secondName: UnsafeMutableRawPointer?
+    
+    if let nodeFirstName = node.firstName {
+      var text = nodeFirstName.text
+      firstName = text.withUTF8 { buf in
+        SwiftASTContext_getIdentifier(ctx, buf.baseAddress, buf.count).raw
+      }
+    } else {
+      firstName = nil
+    }
+    
+    if let nodeSecondName = node.secondName {
+      var text = nodeSecondName.text
+      secondName = text.withUTF8 { buf in
+        SwiftASTContext_getIdentifier(ctx, buf.baseAddress, buf.count).raw
+      }
+    } else {
+      secondName = nil
+    }
+    
+    return ParamDecl_create(ctx, loc, loc, firstName, loc, secondName, declContext)
+  }
+
+  public func visit(_ node: FunctionDeclSyntax) -> UnsafeMutableRawPointer {
+    let loc = self.base.advanced(by: node.position.utf8Offset).raw
+    
+    var nameText = node.identifier.text
+    let name = nameText.withUTF8 { buf in
+      return SwiftASTContext_getIdentifier(ctx, buf.baseAddress, buf.count)
+    }
+    
+    let body: UnsafeMutableRawPointer?
+    if let nodeBody = node.body {
+      body = visit(nodeBody)
+    } else {
+      body = nil
+    }
+
+    let returnType: UnsafeMutableRawPointer?
+    if let output = node.signature.output {
+      returnType = visit(output.returnType)
+    } else {
+      returnType = nil
+    }
+    
+    let params = node.signature.input.parameterList.map { visit($0) }
+    return params.withBridgedArrayRef { ref in
+      FuncDecl_create(ctx, loc, false, loc, name, loc, false, nil, false, nil, loc, ref, loc, body, returnType, declContext)
```

Comment:
I really wish we had argument labels here. Just an idea: You define `FuncDecl_create` on the C++ side and on the Swift side you define 

```swift
struct FuncDecl {
  var pointer: UnsafeMutableRawPointer

  init(<all the parameters>) {
    self.pointer = FuncDecl_create(<forward all the parameters>)
  }
}
```

The downside is that if you modify the parameters on `FuncDecl_create`, you also need to modify them on `FuncDecl.init`. But we could maybe also use `FuncDecl` as a typed wrapper instead of `UnsafeMutableRawPointer` in ASTGen.

And another thought: It might be good to namespace these types in an `enum AST` so they don’t conflict with types from SwiftSyntax or SwiftSyntaxBuilder.